MHArmory.Search/OpenCL/SearchDataSerializer.cs

```diff
@@ -0,0 +1,243 @@
+            }
+
+            byte[] header = new byte[13];
+            SerializeWeaponSlots(data.WeaponSlots, header);
+            header[4] = (byte)heads.Count;
+            header[5] = (byte)chests.Count;
+            header[6] = (byte)gloves.Count;
+            header[7] = (byte)waists.Count;
+            header[8] = (byte)legs.Count;
+            header[9] = (byte)charms.Count;
+            header[10] = (byte)data.AllJewels.Length;
+            header[11] = (byte)data.DesiredAbilities.Length;
+            header[12] = (byte)maps.SetIdMap.Count;
+            parameters.Header = header;
+
+            parameters.Equipment = SerializeEquipment(maps, allEquipment);
+            parameters.DesiredSkills = SerializeDesiredSkills(maps, data.DesiredAbilities);
+            parameters.Decorations = SerializeDecorations(maps, data.AllJewels);
+            parameters.Combinations = (uint) (heads.Count * chests.Count * gloves.Count * waists.Count * legs.Count * charms.Count);
+            parameters.SearchIDMaps = maps;
+            return parameters;
+        }
+
+        private IDictionary<T, byte> CreateIDMap<T>(IEnumerable<T> keys)
+        {
+            byte currentIndex = 0;
+            var map = new Dictionary<T, byte>();
+            foreach (T key in keys)
+            {
+                if (!map.ContainsKey(key))
+                {
+                    map.Add(key, currentIndex);
+                    currentIndex++;
+                }
+            }
+            return map;
+        }
+
+        private void SerializeWeaponSlots(int[] weaponSlots, byte[] header)
+        {
+            foreach (int weaponSlot in weaponSlots)
+            {
+                header[weaponSlot]++;
+            }
+        }
+
+        private byte[] SerializeEquipment(SearchIDMaps maps, IEnumerable<IEquipment> equipments)
+        {
+            var ms = new MemoryStream();
+            var writer = new BinaryWriter(ms);
+            foreach (IEquipment equipment in equipments)
+            {
+                var tuple = new Tuple<int, EquipmentType>(equipment.Id, equipment.Type);
+                byte mappedID = maps.EquipmentIdMap[tuple];
+                writer.Write((ushort)mappedID);
+                writer.Write((ushort)equipment.Id);
+
+                SerializeEquipmentAbilities(maps, equipment, writer);
+                if (equipment is IArmorPiece armorPiece)
+                {
+                    SerializeArmorSetSkills(maps, armorPiece, writer);
+                }
+                else
+                {
+                    for (int i = 0; i < SearchLimits.SetSkillsPerEquipment; i++)
+                    {
+                        SerializeNullSetSkill(writer);
+                    }
+                }
+                byte[] slots = new byte[4];
+                foreach (int slot in equipment.Slots)
+                {
+                    ++slots[slot];
```

Comment:
Please prefer `slots[slot]++` syntax since the other is much less conventional and does not add up anything.